lib/core/Concurrency.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/Concurrency.h>
+
+#include <core/CLogger.h>
+
+#include <boost/make_unique.hpp>
+
+#include <memory>
+#include <thread>
+
+namespace ml {
+namespace core {
+namespace {
+class CExecutorHolder {
+public:
+    CExecutorHolder()
+        : m_ThreadPoolSize{0}, m_Executor{boost::make_unique<transwarp::sequential>()} {}
+
+    static CExecutorHolder makeThreadPool(std::size_t threadPoolSize) {
+        if (threadPoolSize == 0) {
+            threadPoolSize = std::thread::hardware_concurrency();
+        }
+
+        if (threadPoolSize > 0) {
+            try {
+                return CExecutorHolder{threadPoolSize};
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Failed to create thread pool: " << e.what());
+            }
+        } else {
+            LOG_WARN(<< "Failed to determine hardware concurrency");
+        }
+
+        // We'll just have to execute in the main thread.
```

Comment:
do we really want that? Doesn't that break the whole architecture and causes problems?

Maybe failing instead is the better option, otherwise we need to support this case, which also means we have to test it, etc.

Reply:
I mean we could. The thing is there is nothing to stop someone telling the process to run single threaded because of hardware constraints, so we at least need to deal with this case anyway.

Regarding testing, if the program has a bug for a sequential executor it would also have it with a parallel executor, because any sequential run of the program could happen (by chance) when run in parallel. I guess that is the advantage of the executor model: it means we have very little code which differs between the two execution models.